torch/csrc/autograd/variable.h

```diff
@@ -225,10 +222,15 @@ inline const std::shared_ptr<Function>& Variable::grad_fn() const {
   return get()->get_grad_fn();
 };
 inline void Variable::rebase_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn) {
-  get()->requires_grad = flags.requires_grad;
-  get()->is_volatile = flags.is_volatile;
-  get()->output_nr = output_nr;
-  get()->rebase_grad_fn(std::move(grad_fn));
+  if (is_view()) {
+    auto& impl = static_cast<VariableViewImpl&>(*get());
+    impl.rebase_history(flags, output_nr, std::move(grad_fn));
```

Comment:
BTW I don't think this is any faster than a virtual call and it makes the logic slightly more complicated

Reply:
The intention is to make the non-view case fast, by effectively inlining it below. That part can be inlined into the caller, including the `is_view()` check.  

By completely separating the view vs. non-view logic, we're able to check assertions and invariants in VariableViewImpl::rebase_history before we mess with the output_nr or flags.